File: src/main/java/com/uml/generator/model/Trace.java
package com.uml.generator.model;

import com.fasterxml.jackson.annotation.JsonProperty;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * Represents a complete Jaeger trace.
 * A trace contains multiple spans and process information.
 */
public class Trace {

    @JsonProperty("traceID")
    private String traceID;

    @JsonProperty("spans")
    private List<Span> spans;

    @JsonProperty("processes")
    private Map<String, Process> processes;

    @JsonProperty("warnings")
    private List<String> warnings;

    // Non-JSON field: tracks the source filename for this trace
    private String sourceName;

    public Trace() {
        this.spans = new ArrayList<>();
        this.processes = new HashMap<>();
    }

    public String getTraceID() {
        return traceID;
    }

    public void setTraceID(String traceID) {
        this.traceID = traceID;
    }

    public List<Span> getSpans() {
        return spans;
    }

    public void setSpans(List<Span> spans) {
        this.spans = spans;
    }

    public Map<String, Process> getProcesses() {
        return processes;
    }

    public void setProcesses(Map<String, Process> processes) {
        this.processes = processes;
    }

    public List<String> getWarnings() {
        return warnings;
    }

    public void setWarnings(List<String> warnings) {
        this.warnings = warnings;
    }

    public String getSourceName() {
        return sourceName;
    }

    public void setSourceName(String sourceName) {
        this.sourceName = sourceName;
    }

    /**
     * Gets a process by its ID.
     * 
     * @param processID the process ID
     * @return the Process or null if not found
     */
    public Process getProcess(String processID) {
        return processes != null ? processes.get(processID) : null;
    }

    /**
     * Gets the service name for a given span.
     * 
     * @param span the span
     * @return the service name or "unknown"
     */
    public String getServiceName(Span span) {
        if (span == null || span.getProcessID() == null) {
            return "unknown";
        }

        Process process = getProcess(span.getProcessID());
        return process != null ? process.getServiceName() : "unknown";
    }

    /**
     * Gets a span by its ID.
     * 
     * @param spanID the span ID
     * @return the Span or null if not found
     */
    public Span getSpan(String spanID) {
        if (spans == null)
            return null;

        return spans.stream()
                .filter(s -> spanID.equals(s.getSpanID()))
                .findFirst()
                .orElse(null);
    }

    /**
     * Gets all root spans (spans with no parent).
     * 
     * @return list of root spans
     */
    public List<Span> getRootSpans() {
        if (spans == null)
            return new ArrayList<>();

        return spans.stream()
                .filter(Span::isRootSpan)
                .collect(Collectors.toList());
    }

    /**
     * Gets child spans of a given parent span.
     * 
     * @param parentSpanId the parent span ID
     * @return list of child spans
     */
    public List<Span> getChildSpans(String parentSpanId) {
        if (spans == null)
            return new ArrayList<>();

        return spans.stream()
                .filter(s -> parentSpanId.equals(s.getParentSpanId()))
                .collect(Collectors.toList());
    }

    /**
     * Gets all unique service names in this trace.
     * 
     * @return set of service names
     */
    public List<String> getAllServiceNames() {
        if (spans == null || processes == null) {
            return new ArrayList<>();
        }

        return spans.stream()
                .map(span -> getServiceName(span))
                .distinct()
                .sorted()
                .collect(Collectors.toList());
    }

    /**
     * Sorts spans by start time.
     * 
     * @return sorted list of spans
     */
    public List<Span> getSpansSortedByTime() {
        if (spans == null)
            return new ArrayList<>();

        return spans.stream()
                .sorted((s1, s2) -> Long.compare(s1.getStartTime(), s2.getStartTime()))
                .collect(Collectors.toList());
    }

    @Override
    public String toString() {
        return "Trace{" +
                "traceID='" + traceID + '\'' +
                ", spans=" + (spans != null ? spans.size() : 0) +
                ", processes=" + (processes != null ? processes.size() : 0) +
                '}';
    }
}

File: python_version/jaeger_uml_generator/models/trace.py
"""Trace model representing a complete distributed trace."""

from dataclasses import dataclass, field
from typing import List, Dict, Optional
from .span import Span
from .process import Process


@dataclass
class Trace:
    """Represents a complete Jaeger trace with spans and processes."""
    
    trace_id: str
    spans: List[Span] = field(default_factory=list)
    processes: Dict[str, Process] = field(default_factory=dict)
    warnings: List[str] = field(default_factory=list)
    source_name: Optional[str] = None
    
    @classmethod
    def from_dict(cls, data: dict) -> 'Trace':
        """Create a Trace from a dictionary."""
        # Parse spans
        spans = []
        if 'spans' in data and isinstance(data['spans'], list):
            spans = [Span.from_dict(span_data) for span_data in data['spans']]
        
        # Parse processes
        processes = {}
        if 'processes' in data and isinstance(data['processes'], dict):
            for process_id, process_data in data['processes'].items():
                processes[process_id] = Process.from_dict(process_data)
        
        warnings = data.get('warnings', [])
        
        return cls(
            trace_id=data.get('traceID', ''),
            spans=spans,
            processes=processes,
            warnings=warnings
        )
    
    def get_process(self, process_id: str) -> Optional[Process]:
        """Get a process by its ID."""
        return self.processes.get(process_id)
    
    def get_service_name(self, span: Span) -> str:
        """Get the service name for a given span."""
        if not span or not span.process_id:
            return 'unknown'
        
        process = self.get_process(span.process_id)
        return process.service_name if process else 'unknown'
    
    def get_span(self, span_id: str) -> Optional[Span]:
        """Get a span by its ID."""
        for span in self.spans:
            if span.span_id == span_id:
                return span
        return None
    
    def get_root_spans(self) -> List[Span]:
        """Get all root spans (spans with no parent)."""
        return [span for span in self.spans if span.is_root_span()]
    
    def get_child_spans(self, parent_span_id: str) -> List[Span]:
        """Get child spans of a given parent span."""
        children = []
        for span in self.spans:
            if span.get_parent_span_id() == parent_span_id:
                children.append(span)
        return children
    
    def get_all_service_names(self) -> List[str]:
        """Get all unique service names in this trace, sorted."""
        service_names = set()
        for span in self.spans:
            service_name = self.get_service_name(span)
            service_names.add(service_name)
        return sorted(service_names)
    
    def get_spans_sorted_by_time(self) -> List[Span]:
        """Get spans sorted by start time."""
        return sorted(self.spans, key=lambda s: s.start_time)
    
    def __str__(self) -> str:
        return f"Trace(trace_id='{self.trace_id}', spans={len(self.spans)}, processes={len(self.processes)})"
